### PR TITLE
Document adding notification modes

### PR DESCRIPTION
**Pull request by trasher**
Document adding notifications modes (see https://github.com/glpi-project/glpi/pull/2207). Should not be merged before the PR on the core has been as well.

View online:
http://glpi-developer-documentation.readthedocs.io/en/feature-notifs_modes/plugins/notifications.html

### CODE REVIEW

**Comment by trasher** [comment]
Note: links to the apidoc references classes that does not exists yet; and are therefore broken.

File: source/plugins/notifications.rst [hunk 1]
```diff
@@ -0,0 +1,334 @@
+Notification modes
+------------------
+
+Core GLPI provides two notifications modes as of today:
+
+* email (sends email),
+* ajax (send brower notifications if/when user is logged)
+
+It is possible to extends this mecanism in order to create another mode to use. Let's take a tour... We'll take example of a plugin designed to send SMS to the users.
```
Comment (orthagh) [inline source/plugins/notifications.rst]: mechanism

File: source/plugins/notifications.rst [hunk 2]
```diff
@@ -0,0 +1,334 @@
+Notification modes
+------------------
+
+Core GLPI provides two notifications modes as of today:
+
+* email (sends email),
+* ajax (send brower notifications if/when user is logged)
```
Comment (orthagh) [inline source/plugins/notifications.rst]: browser

File: source/plugins/notifications.rst [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+Event
+^^^^^
+
+Once the new mode has been enabled; it will try to raise core events. You will need to create an event class named ``PluginSmsNotificationEventSms`` that implements `NotificationEventInterface <https://forge.glpi-project.org/apidoc/class-NotificationEventInterface.html>`_ and extends `NotificationEventAbstract <https://forge.glpi-project.org/apidoc/class-NotificationEventAbstract.html>` in the ``inc/notifictioneventsms.php``.
+
+Methods to implement are:
+
+* ``getTargetFieldName``: defines the name of the target field;
+* ``getTargetField``: populate if needed the target field to use. For a SMS plugin, it would retrieve the phone number from users table for example;
+* ``canCron``: whether notification can be called from a contask. For the SMS plugins, it would be true. It is set to false for ajax based events; because notifications are requested from user browser;
```
Comment (orthagh) [inline source/plugins/notifications.rst]: crontask

File: source/plugins/notifications.rst [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+* ``getAdminData``: as global admin is not a real user; you can define here the data used to send the notification;
+* ``getEntityAdminData``: same as the above, but for entities admins rather than global admin;
+* ``send``: method that will really send data.
+
+The ``raise`` method declared in the interface is implemented in the abstract class; since it should be used as it for every mode. If you want to do extra process in the ``raise`` method, you should override the ``extraRaise`` method. This is done in the core to add signatures in the mailing for example.
+
+.. note::
+
+   Notifications uses the ``QueueNotification`` to store its data. Each notificfation about to be sent will be stored in the relevant table. Rows are updated once the notification has really be send (set ``is_deleted`` to 1 and update ``sent_time``.
```
Comment (orthagh) [inline source/plugins/notifications.rst]: notification

File: source/plugins/notifications.rst [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+         'server' => '',
+         'port'   => ''
+      ]
+   );
+
+   //get configuration
+   $conf = Config::getConfigurationValues('plugin:sms');
+   //$conf will be ['server' => '', 'port' => '']
+
+That said, we need to create a class to handle the settings, and a front file to display them. The class must be named ``PluginSmsNotificationSmsSetting`` and must be in the ``inc/notifictionsmssetting.class.php``. It have to extends the `NotificationSetting <https://forge.glpi-project.org/apidoc/class-NotificationSetting.html>`_ core class :
```
Comment (orthagh) [inline source/plugins/notifications.rst]: notificationsmssetting

File: source/plugins/notifications.rst [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+         }
+
+         //Ignore display parameter since showFormButtons is now ready :/ (from all but tests)
+         echo $out;
+
+         $this->showFormButtons($options);
+      }
+   }
+
+The front form file, located at ``front/notifictionsmssetting.form.php`` will be quite simple. It handles the display of the configuration form, update of the values, and test send (if any):
```
Comment (orthagh) [inline source/plugins/notifications.rst]: notificationsmssetting

File: source/plugins/notifications.rst [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+   Html::header(Notification::getTypeName(Session::getPluralNumber()), $_SERVER['PHP_SELF'], "config", "notification", "config");
+
+   $notificationsms->display(array('id' => 1));
+
+   Html::footer();
+
+Event
+^^^^^
+
+Once the new mode has been enabled; it will try to raise core events. You will need to create an event class named ``PluginSmsNotificationEventSms`` that implements `NotificationEventInterface <https://forge.glpi-project.org/apidoc/class-NotificationEventInterface.html>`_ and extends `NotificationEventAbstract <https://forge.glpi-project.org/apidoc/class-NotificationEventAbstract.html>` in the ``inc/notifictioneventsms.php``.
```
Comment (orthagh) [inline source/plugins/notifications.rst]: notificationsmssetting

File: source/plugins/notifications.rst [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+
+Methods to implement are:
+
+* ``getTargetFieldName``: defines the name of the target field;
+* ``getTargetField``: populate if needed the target field to use. For a SMS plugin, it would retrieve the phone number from users table for example;
```
Comment (orthagh) [inline source/plugins/notifications.rst]: populates ? (i'm not sure)

**Pull request merged**